sklearn/utils/_param_validation.py

```diff
@@ -0,0 +1,391 @@
+from abc import ABC
+from abc import abstractmethod
+import functools
+from inspect import signature
+from numbers import Integral
+from numbers import Real
+import operator
+
+import numpy as np
+from scipy.sparse import issparse
+
+from .validation import _is_arraylike_not_scalar
+
+
+def validate_parameter_constraints(parameter_constraints, params):
+    """Validate types and values of given parameters.
+
+    Parameters
+    ----------
+    parameter_constraints : dict
+        A dictionary `param_name: list of constraints`. A parameter is valid if it
+        satisfies one of the constraints from the list. Constraints can be:
+        - an Interval object, representing a continuous or discrete range of numbers
+        - the string "array-like"
+        - the string "sparse matrix"
+        - callable
+        - None, meaning that None is a valid value for the parameter
+        - any type, meaning that any instance of this type is valid
+        - a StrOptions object, representing a set of strings
```

Comment:
Some interfaces parameters are valid if provided as unique value. This is for instance the case for `n_jobs` with `-1`.

Do we want to encode such cases as a `_Constraint`?

Reply:
Regarding special constraints, there are constraints that are not implemented here like: being an estimator instance, being a CV instance, ...
I think it's better to define new constraint objects as we find new needs, instead of trying to find all special cases in advance and add unused code that will probably be bugged.